TypeCobol.LanguageServer/Workspace.cs

```diff
@@ -630,42 +726,69 @@ public void DidChangeConfigurationParams(string[] arguments)
+            this.WorkspaceProjectStore.RefreshOpenedFiles();
+        }
 
-            lock (_lockForOpenedDocuments)
-            {
-                var tempOpenedContexts = new Dictionary<Uri, DocumentContext >(_openedDocuments);
-                foreach (var docContext in tempOpenedContexts.Values)
-                {
-                    var compilationResults = docContext.FileCompiler?.CompilationResultsForProgram;
-                    if (compilationResults == null) continue;
+        /// <summary>
+        /// Refresh One opened document
+        /// </summary>
+        /// <param name="docContext">The Document context to be refreshed</param>
+        /// <param name="clearCopyCache">true if the imported document cache must be cleared, false otherwise</param>
+        private void RefreshOpenedDocument(DocumentContext docContext, bool clearCopyCache)
+        {
+            if (clearCopyCache)
+                docContext.FileCompiler.CompilationProject.ClearImportedCompilationDocumentsCache();
 
-                    var sourceText = new StringBuilder();
-                    foreach (var cobolTextLine in compilationResults.CobolTextLines)
-                        sourceText.AppendLine(cobolTextLine.Text);
+            var sourceText = new StringBuilder();
+            var compilationResults = docContext.FileCompiler?.CompilationResultsForProgram;
+            if (compilationResults == null)
+                return;
+            foreach (var cobolTextLine in compilationResults.CobolTextLines)
+                sourceText.AppendLine(cobolTextLine.Text);
 
-                    //Disconnect previous LanguageServer connection
-                    docContext.LanguageServerConnection(false);                    
+            //Disconnect previous LanguageServer connection
+            docContext.LanguageServerConnection(false);
 
-                    OpenTextDocument(docContext, sourceText.ToString(), LsrTestingOptions.NoLsrTesting);
-                }
+            BindFileCompilerSourceTextDocument(docContext, sourceText.ToString(), LsrTestingOptions.NoLsrTesting);
+        }
+
+        /// <summary>
+        /// Refresh all documents of a WorkspaceProject instance
+        /// </summary>
+        internal void ScheduleRefresh(WorkspaceProject project, bool clearCopyCache)
+        {
+            if (clearCopyCache)
+            {
+                project.Project.ClearImportedCompilationDocumentsCache();
+            }
+            if (!project.IsEmpty)
+            {
+                Action action = () =>
+                {
+                    foreach (var docContext in project.OpenedDocuments)
+                    {
+                        System.Diagnostics.Debug.Assert(docContext.FileCompiler.CompilationProject == project.Project);
+                        RefreshOpenedDocument(docContext, false);
+                    }
+                };
+                MessagesActionsQueue.Enqueue(new MessageActionWrapper(action));
```

Comment:
I was wrong in #2178, there is no need to wrap the refresh into an action on the message queue. We are already on the correct thread here so you can simplify this to:

```csharp
foreach (var docContext in project.OpenedDocuments)
{
    System.Diagnostics.Debug.Assert(docContext.FileCompiler.CompilationProject == project.Project);
    RefreshOpenedDocument(docContext, false);
}
```

Sorry for the incorrect suggestion.

Reply:
changed in [8d9c2da](https://github.com/TypeCobolTeam/TypeCobol/pull/2158/commits/8d9c2daba074a18dbf5260d351d3278d470ebf44)